statemachine/statemachine.py

```diff
@@ -140,35 +143,20 @@ def _visit_states_and_transitions(self, visitor):
             for transition in state.transitions:
                 visitor(transition)
 
-    def _setup(self, class_states: States, initial_transition: Transition):
+    def _setup(self, initial_transition: Transition):
         """
         Args:
-            class_states: The original (shared) instances of :ref:`State` living on the
-                user's concrete :ref:`StateMachine` class. These instances cannot be modified.
-                So we will clone the states in order to bind the new instances with concrete
-                actions and event handlers.
-
             initial_transition: A special :ref:`transition` that triggers the enter on the
                 `initial` :ref:`State`.
         """
-        machine = ObjectConfig(self, skip_attrs=self._get_protected_attrs())
-        model = ObjectConfig(self.model, skip_attrs={self.state_field})
+        machine = ObjectConfig.from_obj(self, skip_attrs=self._get_protected_attrs())
+        model = ObjectConfig.from_obj(self.model, skip_attrs={self.state_field})
         default_resolver = resolver_factory(machine, model)
 
-        # clone states and transitions to avoid sharing callbacks references between instances
-        self.states = States(
-            {s.id: s.clone()._setup(self, default_resolver) for s in class_states}
```

Comment:
This is the main change on this PR: Instead of cloning states to get isolated SM instances, we now register the callbacks references on the SM instance. So States and Transitions are kind of immutable at SM runtime.